Mathlib/Algebra/Star/Positive.lean

```diff
@@ -0,0 +1,187 @@
+* `StarOrderedRing.positive.selfAdjoint : positive R →+ selfAdjoint R`: the identity map as an
+  `AddMonoidHom`.
+* `StarOrderedRing.positive.instAddMonoidWithOne`: Makes `positive R` into an `AddMonoidWithOne`
+* `StarOrderedRing.positive.instCanonicallyOrderedAddMonoid`: Makes `positive R` into a
+  `CanonicallyOrderedAddMonoid` which avoids duplication of some lemmas (e.g., `zero_le`)
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+namespace StarOrderedRing
+
+/-- The positive elements of a star ordered ring, as an additive submonoid.
+
+Note that the type class assumptions for `StarOrderedRing.positive` are actually significantly
+weaker (just `OrderedAddCommMonoid`) so that it can be used in as many contexts as possible, but
+the primary intended use case is when `R` is a `StarOrderedRing`.
+
+In addiiton, we note that `StarOrderedRing.positive` is just an alias of `AddSubmonoid.nonneg`, but
+we choose to keep these separate for naming purposes and also because we will add many instances to
+this object (viewed as a type).
+
+Regarding naming of lemmas: whenever a lemma involves a hypothesis of the form `0 ≤ x`, it will be
+referenced as `nonneg` in the declaration name, and we always prefer `0 ≤ x` over `x ∈ positive R`,
+even though these are definitionally equal. However, when the hypothesis is `x : positive R` (note
+the `:` versus `∈`), the declaration name will use `positive`, not `pos`, so as to avoid confusion
+with `0 < x`. -/
+alias AddSubmonoid.nonneg ← positive
+
+theorem positive_def [OrderedAddCommMonoid R] {x : R} :
+    x ∈ positive R ↔ 0 ≤ x :=
+  Iff.rfl
+
+theorem positive_coe_set [OrderedAddCommMonoid R] : (positive R : Set R) = {x | 0 ≤ x} := rfl
+
+section NonUnitalSemiring
+
+variable [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+
+theorem positive_iff {x : R} :
+    x ∈ positive R ↔ x ∈ AddSubmonoid.closure (Set.range fun s : R => star s * s) := by
+  rw [positive_def, nonneg_iff]
+
+theorem isSelfAdjoint_of_nonneg [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+    {x : R} (hx : 0 ≤ x) : IsSelfAdjoint x := by
+  rw [nonneg_iff] at hx
+  refine AddSubmonoid.closure_induction hx ?_ (isSelfAdjoint_zero R) fun _ _ => IsSelfAdjoint.add
+  rintro - ⟨s, rfl⟩
+  exact IsSelfAdjoint.star_mul_self s
+
+theorem le_iff_exists_positive {x y : R} : x ≤ y ↔ ∃ p : positive R, y = x + p := by
+  rw [le_iff']
+  simp only [Subtype.exists, exists_prop]
+  rfl
+
+@[norm_cast]
+theorem positive.val_le_val {x y : positive R} : (x : R) ≤ y ↔ x ≤ y :=
+  Iff.rfl
+
+instance positive.instCanonicallyOrderedAddMonoid : CanonicallyOrderedAddMonoid (positive R) :=
+  { inferInstanceAs (OrderedAddCommMonoid (positive R)) with
+    bot := 0
+    bot_le := Subtype.prop
+    exists_add_of_le := @fun x y (h : (x : R) ≤ y) => by
+      convert le_iff_exists_positive.mp h using 1
+      funext p
+      exact_mod_cast rfl
+    le_self_add := fun x y => show (x : R) ≤ x + y from le_iff_exists_positive.mpr ⟨y, rfl⟩ }
```

Comment:
Isn't this just [`Nonneg.canonicallyOrderedAddMonoid`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Order/Nonneg/Ring.html#Nonneg.canonicallyOrderedAddMonoid)?

Reply:
No, that requires an `OrderedRing` instance (although I didn't actually check if it's necessary).